components/supervisor/pkg/ports/ports-config.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package ports
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/gitpod-io/gitpod/supervisor/pkg/gitpod"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// RangeConfig is a port range config
+type RangeConfig struct {
+	*gitpod.PortsItems
+	Start uint32
+	End   uint32
+}
+
+// Configs provides access to port configurations
+type Configs struct {
+	workspaceConfigs     map[uint32]*gitpod.PortConfig
+	instancePortConfigs  map[uint32]*gitpod.PortConfig
+	instanceRangeConfigs []*RangeConfig
+}
+
+// ForEach iterates over all configured ports
+func (configs *Configs) ForEach(callback func(port uint32, config *gitpod.PortConfig)) {
+	if configs == nil {
+		return
+	}
+	visited := make(map[uint32]struct{})
+	for _, configs := range []map[uint32]*gitpod.PortConfig{configs.instancePortConfigs, configs.workspaceConfigs} {
+		for port, config := range configs {
+			_, exists := visited[port]
+			if exists {
+				continue
+			}
+			visited[port] = struct{}{}
+			callback(port, config)
+		}
+	}
+}
+
+// Get returns the config for the give port
+func (configs *Configs) Get(port uint32) (*gitpod.PortConfig, bool) {
+	if configs == nil {
+		return nil, false
+	}
+	config, exists := configs.instancePortConfigs[port]
+	if exists {
+		return config, true
+	}
+	config, exists = configs.workspaceConfigs[port]
+	if exists {
+		return config, true
+	}
+	return nil, false
+}
+
+// GetFromRange returns the port config from the port range config
+func (configs *Configs) GetFromRange(port uint32) (*gitpod.PortConfig, bool) {
```

Comment:
Does it matter to the config consumer if the config comes from an individual bit of config or from a port range?
Looking at the callsites, it seems that this design forces consumers to know about the distinction and make a second call if need be.

If the distinction doesn't matter, this bit could go into `Get`.
If the distinction does matter, `Get` could return where the config comes from.